security-analysis/security-analysis-api/src/test/java/com/powsybl/security/SecurityAnalysisResultBuilderTest.java

```diff
@@ -33,29 +37,46 @@ public void failedResult() {
 
     @Test
     public void completeResult() {
+        SecurityAnalysisInterceptor securityAnalysisInterceptorMock = new MockInterceptor();
+        MockViolationContextExt preVioContextExt = new MockViolationContextExt();
 
         SecurityAnalysisResultBuilder builder = new SecurityAnalysisResultBuilder(new LimitViolationFilter(),
-                new RunningContext(network, network.getVariantManager().getWorkingVariantId()));
+                new RunningContext(network, network.getVariantManager().getWorkingVariantId()), Collections.singleton(securityAnalysisInterceptorMock));
 
         VoltageLevel vl = network.getVoltageLevel("VLHV1");
         vl.getBusView().getBusStream().forEach(b -> b.setV(410));
 
+        SecurityAnalysisResultBuilder.PreContingencyResultBuilder preContingencyResultBuilder = builder.preContingency();
+        SecurityAnalysisResultBuilder securityAnalysisResultBuilder = preContingencyResultBuilder.setComputationOk(true)
+                .addViolationContextExtension(Security.checkLimits(network).get(0), preVioContextExt)
```

Comment:
From my point of view, this is not a good design. With this one, we fill the SecurityAnalysisResultBuilder with a lot of information that are used only once.

The use case is:
- the SecurityAnalysisResultBuilder is responsible on creating the contingencies and the violations.
- the SecurityAnalysisInterceptors are responsible on adding the additional data to the contingencies and the violations

The SecurityAnalysisInterceptors need a way to access to the network or to another data source (the result of the simulation). I think we should keep only the RunningContext with extensions. However, the Interceptors should have a new notification method to distinguish the violation creation in pre and post contingency state:
```
public void onLimitViolation(RunningContext, Contingency contingency, LimitViolation violation);
```

With the contingency name, and the violation, the extension in the running context should be able to retrieve the relevant information from an additional datasource. In that case, I'm not convinced we should keep ViolationContext and ContingencyContext, but only RunningContext, that is global to the simulation. Separate everything in really small classes make the API hard to understand.

Reply:
The design with more specific `ViolationContext` and `ContingencyContext` comes from a comment from @jonenst about not changing the currently immutable class. If we keep only one `RunningContext` but make it very "mutable" by adding and removing extensions, it will not be possible to ensure its consistence, in particular in case of concurrent use.

I think we could keep only the `RunningContext` class as you propose, but in that case it's important to create a new one for each notification of the interceptor, so that it stays thread safe.

Reply:
Then, don't make the RunningContext mutable using the extension mechanism and create subclasses? I think creating a lot of small objects with very short life cycle is not a good idea. You can also create a decorator to hide methods you don't want someone use (probably removing an extension). There is a lot of way to solve this, isn't it?

Reply:
But then again you leave the implentation deal with a big unique stateful object, which is not convenient for thread safety, and may prevent to use the result builder "on the go": you will need to read all the result in order to be able to create your context.

It will not either be more performant I think, because you will probably need to create maps and search in them frequently, whereas small objects carry just the information you need.